nexus/db-queries/src/db/datastore/vmm.rs

```diff
@@ -164,4 +172,52 @@ impl DataStore {
 
         Ok(vmm)
     }
+
+    /// Lists VMMs which have been abandoned by their instances after a
+    /// migration and are in need of cleanup.
+    ///
+    /// A VMM is considered "abandoned" if (and only if):
+    ///
+    /// - It is in the `Destroyed` state.
+    /// - It has previously been asigned to an instance.
+    /// - It is not currently running the instance, and it is also not the
+    ///   migration target of that instance (i.e. it is no longer pointed to by
+    ///   the instance record's `active_propolis_id` and `target_propolis_id`
+    ///   fields).
+    /// - It has not been deleted yet.
+    pub async fn vmm_list_abandoned(
+        &self,
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, Uuid>,
+    ) -> ListResultVec<Vmm> {
+        use crate::db::schema::instance::dsl as instance_dsl;
+        let destroyed = DbInstanceState::new(ApiInstanceState::Destroyed);
+        paginated(dsl::vmm, dsl::id, pagparams)
+            // In order to be considered "abandoned", a VMM must be:
+            // - in the `Destroyed` state
+            .filter(dsl::state.eq(destroyed))
+            // - not deleted yet
+            .filter(dsl::time_deleted.is_null())
+            // - not pointed to by their corresponding instances
+            .left_join(
```

Comment:
(as mentioned in the hypervisor meeting) the fact that this is a `left_join`, not an `inner_join`, I *think* means we would also find any "VMMs that are 'destroyed', but not 'deleted', as long as instances aren't pointing at them".

I think this is actually a nice-to-have feature -- could (should?) this background task be responsible for destroying VMMs that are in a terminal state?

Reply:
Yeah, we discussed this during the hypervisor sync --- I've updated the comments for this to more accurately reflect that this task will also clean up VMMs that are `Destroyed` and have *never* been assigned to an instance. The one case that this task is _not_ responsible for is for VMMs that are `Destroyed`/`Failed` while being used actively to run an instance, which is the responsibility of the (future) `instance-update` saga.

Reply:
Sounds good - I think I was a bit narrowly focused on "what is the data", rather than "how did we get in this state" in my review, but your comments here make a lot of sense, especially in the context of the upcoming `instance-update` saga.